awx/main/wsbroadcast.py

```diff
@@ -0,0 +1,190 @@
+                                             timeout=timeout) as session:
+                async with session.ws_connect(uri, ssl=self.verify_ssl) as websocket:
+                    self.stats.record_connection_established()
+                    attempt = 0
+                    await self.run_loop(websocket)
+        except asyncio.CancelledError:
+            # TODO: Check if connected and disconnect
+            # Possibly use run_until_complete() if disconnect is async
+            logger.warn(f"{self.name} connection to {self.remote_host} cancelled")
+            self.stats.record_connection_lost()
+            raise
+        except client_exceptions.ClientConnectorError as e:
+            logger.warn(f"Failed to connect to {self.remote_host}: '{e}'. Reconnecting ...")
+            self.stats.record_connection_lost()
+            self.start(attempt=attempt + 1)
+        except asyncio.TimeoutError:
+            logger.warn(f"Timeout while trying to connect to {self.remote_host}. Reconnecting ...")
+            self.stats.record_connection_lost()
+            self.start(attempt=attempt + 1)
+        except Exception as e:
+            # Early on, this is our canary. I'm not sure what exceptions we can really encounter.
+            logger.warn(f"Websocket broadcast client exception {type(e)} {e}")
+            self.stats.record_connection_lost()
+            self.start(attempt=attempt + 1)
+
+    def start(self, attempt=0):
+        self.async_task = self.event_loop.create_task(self.connect(attempt=attempt))
+
+    def cancel(self):
+        self.async_task.cancel()
+
+
+class BroadcastWebsocketTask(WebsocketTask):
+    async def run_loop(self, websocket: aiohttp.ClientWebSocketResponse):
+        async for msg in websocket:
+            self.stats.record_message_received()
+
+            if msg.type == aiohttp.WSMsgType.ERROR:
+                break
+            elif msg.type == aiohttp.WSMsgType.TEXT:
+                try:
+                    payload = json.loads(msg.data)
+                except json.JSONDecodeError:
+                    logmsg = "Failed to decode broadcast message"
+                    if logger.isEnabledFor(logging.DEBUG):
+                        logmsg = "{} {}".format(logmsg, payload)
+                    logger.warn(logmsg)
+                    continue
+
+                (group, message) = unwrap_broadcast_msg(payload)
+
+                await self.channel_layer.group_send(group, {"type": "internal.message", "text": message})
+
+
+class BroadcastWebsocketManager(object):
+    def __init__(self):
+        self.event_loop = asyncio.get_event_loop()
+        self.broadcast_tasks = dict()
+        # parallel dict to broadcast_tasks that tracks stats
+        self.local_hostname = get_local_host()
+        self.stats_mgr = BroadcastWebsocketStatsManager(self.event_loop, self.local_hostname)
+
+    async def run_per_host_websocket(self):
+
+        while True:
+            future_remote_hosts = get_broadcast_hosts()
```

Comment:
Okay, so this makes sense to me.  Every 10 seconds, look for new or cluster peers, and either cancel and stop tracking them, or begin broadcasting websocket messages to them.